src/query/method_def.cpp

```diff
@@ -0,0 +1,146 @@
+/*
+ *
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#include "method_def.hpp"
+
+#include "memory_private_allocator.hpp"
+
+void
+method_sig_list::freemem ()
+{
+  METHOD_SIG *sig = method_sig;
+  while (sig)
+    {
+      METHOD_SIG *next = sig->next;
+
+      db_private_free_and_init (NULL, sig->method_name);
+      if (sig->class_name)
+	{
+	  db_private_free_and_init (NULL, sig->class_name);
+	}
+      db_private_free_and_init (NULL, sig->method_arg_pos);
+      db_private_free_and_init (NULL, sig); /* itself */
+
+      sig = next;
+    }
+}
+
+void
+method_sig_list::pack (cubpacking::packer &serializator) const
+{
+  serializator.pack_int (num_methods);
+
+  METHOD_SIG *sig_p = method_sig;
+  while (sig_p)
+    {
+      serializator.pack_c_string (sig_p->method_name, strlen (sig_p->method_name));
+
+      serializator.pack_int (sig_p->method_type);
+      serializator.pack_int (sig_p->num_method_args);
+
+      for (int i = 0; i < sig_p->num_method_args + 1; i++)
+	{
+	  serializator.pack_int (sig_p->method_arg_pos[i]);
+	}
+
+      if (sig_p->method_type != METHOD_IS_JAVA_SP)
+	{
+	  serializator.pack_c_string (sig_p->class_name, strlen (sig_p->class_name));
+	}
+
+      sig_p = sig_p->next;
+    }
+}
+
+size_t
+method_sig_list::get_packed_size (cubpacking::packer &serializator, std::size_t start_offset) const
+{
+  size_t size = serializator.get_packed_int_size (start_offset); /* num_methods */
+
+  METHOD_SIG *sig_p = method_sig;
+  while (sig_p)
+    {
+      size += serializator.get_packed_c_string_size (sig_p->method_name, strlen (sig_p->method_name), size);
+
+      /* method type and num_method_args */
+      size += serializator.get_packed_int_size (size);
+      size += serializator.get_packed_int_size (size);
+
+      for (int i = 0; i < sig_p->num_method_args + 1; i++)
+	{
+	  size += serializator.get_packed_int_size (size); /* method_sig->method_arg_pos[i] */
+	}
+
+      if (sig_p->method_type != METHOD_IS_JAVA_SP)
+	{
+	  size += serializator.get_packed_c_string_size (sig_p->class_name, strlen (sig_p->class_name), size);
+	}
+
+      sig_p = sig_p->next;
+    }
+  return size;
+}
+
+void
+method_sig_list::unpack (cubpacking::unpacker &deserializator)
+{
+  // assuming that METHOD_SIG_LIST is already allocated
+
+  deserializator.unpack_int (num_methods);
```

Comment:
not needed to check if num_methods is 0

Reply:
it is an unpacking method, `num_methods` value is set from the buffer